source/code/plugins/oms_common.rb

```diff
@@ -1019,4 +1019,191 @@ def to_s
+      self.instance.cleanup
+    end
+
+    def get_mpid
+      return @master_process
+    end
+
+    def cleanup
+      return if @proc_cache.empty?
+      log "Cleanup jobs, pid=#{Process.pid} mpid=#{self.get_mpid} ppid=#{Process.ppid}"
+
+      @proc_cache.each do |pid, val|
+        Process.kill('SIGTERM', pid)
+      end
+
+      @proc_cache.each do |pid, val|
+        Process.kill('SIGKILL', pid)
+      end
+      @proc_cache_lock.synchronize {
+        @proc_cache.clear
+      }
+    end
+
+    def run_job_and_wait(&block)
+      read_io, write_io = IO.pipe
+
+      # read_s, write_s, = IO.select([read_io], [write_io])
+      # read_io = read_s[0]
+      # write_io = write_s[0]
+
+      pid = fork do
+        ["SIGHUP", "SIGTERM"].each do |sig|
+          Signal.trap(sig) { log "Child process ##{Process.pid} receiving #{sig}\n"; exit }
+        end
+
+        read_io.close # For parent's use, not child's use
+        result = {}
+        begin
+          yield_ret = yield
+          yield_results = {}
+          trace "yield_ret=#{yield_ret}"
+          if yield_ret.is_a?(Array)
+            yield_ret.each { |data|
+              operation, source, event = 'UNKOWN', 'UNKOWN', nil
+              event = data[:event] if data.is_a?(Hash) && data.key?(:event)
+              # event = event[0] if event.is_a?(Array)
+              source = data[:source] if data.is_a?(Hash) && data.key?(:source)
+              if event.is_a?(String)
+                operation = LOG_ERROR
+              elsif event.is_a?(Hash) && event.key?(:op)
+                operation = event[:op]
+              end
+
+              yield_results[:telemetry] = [] unless yield_results.key?(:telemetry)
+              yield_results[:telemetry].push({'operation': operation, 'event': event, 'source': source})
+            }
+          end
+          result[:return] = yield_results
+        rescue => e # We should catch any exception to pass it to parent
+          result[:exception] = {'class': e.class.name, 'msg': e.message, 'backtrace': e.backtrace}
+        end
+        # process is orphan
+        Process.exit(false) if Process.ppid == 1
+
+        write_io.write(result)
+        trace "write_io <= #{result}"
+      end
+
+      add_process_to_cache(pid)
+      write_io.close # For child use
+      # blocking read on pipe
+      results = eval(read_io.read)
+      trace "Receiving results=#{results}"
+
+      read_io.close
+      Process.waitpid(pid)
+      remove_process_from_cache(pid)
+
+      unless results.is_a?(Hash)
+        log "results is not a hash, results=#{results}"
+        return results
+      end
+
+      # Handling Exception
+      if results.key?(:exception) and results[:exception] != nil
+        ex_class_name = results[:exception][:class]
+        ex_msg = results[:exception][:msg]
+        ex_backtrace = results[:exception][:backtrace]
+        if Object.const_defined?(ex_class_name)
```

Comment:
this requires DSC module release

Reply:
Yes, I did some testing with shell bundle 1.9, and it was working.